chain/chain/src/runtime/mod.rs

```diff
@@ -407,18 +406,42 @@ impl NightshadeRuntime {
             part_id,
             trie_with_state,
         );
-        let state_part = borsh::to_vec(&match trie_nodes {
+        let partial_state = match trie_nodes {
             Ok(partial_state) => partial_state,
             Err(err) => {
                 error!(target: "runtime", ?err, part_id.idx, part_id.total, %prev_hash, %state_root, %shard_id, "Can't get trie nodes for state part");
                 return Err(err.into());
             }
-        })
-            .expect("serializer should not fail");
-
+        };
+        let protocol_version = self.epoch_manager.get_epoch_protocol_version(&epoch_id)?;
+        let state_part = StatePart::from_partial_state(partial_state, protocol_version);
         Ok(state_part)
     }
 
+    fn validate_state_part_impl(
+        &self,
+        state_root: &StateRoot,
+        part_id: PartId,
+        part: &StatePart,
+    ) -> bool {
+        let partial_state = match part.to_partial_state() {
+            Ok(partial_state) => partial_state,
+            // Deserialization error means we've got the data from malicious peer
+            Err(err) => {
+                tracing::error!(target: "state-parts", ?err, "State part deserialization error");
+                return false;
+            }
+        };
```

Comment:
nit: What do you think about the alternative below ? I'm not sure myself which is better.
```
        let partial_state = part.to_partial_state();
        let Ok(partial_state) = part.to_partial_state() else {
            tracing::error!(target: "state-parts", ?partial_state, "State part deserialization error");
            return false;
        };
```